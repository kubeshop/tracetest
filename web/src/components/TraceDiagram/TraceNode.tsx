import Text from 'antd/lib/typography/Text';
import {upperCase} from 'lodash';
import React from 'react';
import {Handle, NodeProps, Position} from 'react-flow-renderer';
import {SemanticGroupNames, SemanticGroupNamesToText} from '../../constants/SemanticGroupNames.constants';
import {getSpanNodeInfo} from '../../services/Span.service';
import {ISpan} from '../../types/Span.types';
import {ITrace} from '../../types/Trace.types';
import * as S from './TraceDiagram.styled';

type TTraceNodeProps = NodeProps<{span: ISpan; trace: ITrace}>;

const TraceNode: React.FC<TTraceNodeProps> = ({
  id,
  data: {
    span: {name, spanId},
    trace,
  },
  selected,
}) => {
  const {heading, spanType = SemanticGroupNames.General, primary} = getSpanNodeInfo(spanId, trace);
  const spanTypeText = SemanticGroupNamesToText[spanType];

  return (
    <S.TraceNode selected={selected}>
      <S.TraceNotch spanType={spanType}>
        <Text>{upperCase(heading || spanTypeText)}</Text>
      </S.TraceNotch>
      <Handle type="target" id={id} position={Position.Top} style={{top: 0, borderRadius: 0, visibility: 'hidden'}} />
      <S.TextContainer>
        {primary && <S.NameText strong>{primary}</S.NameText>}
        <S.NameText>{name}</S.NameText>
      </S.TextContainer>
      <Handle
        type="source"
        position={Position.Bottom}
        id={id}
        style={{bottom: 0, borderRadius: 0, visibility: 'hidden'}}
      />
    </S.TraceNode>
  );
};

export default TraceNode;
