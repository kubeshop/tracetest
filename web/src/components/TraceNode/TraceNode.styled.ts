import {Typography} from 'antd';
import styled from 'styled-components';
import {SemanticGroupNames} from '../../constants/SemanticGroupNames.constants';

enum NotchColor {
  HTTP = '#C1E095',
  DB = '#EFDBFF',
  RPC = '#9AD4D6',
  MESSAGING = '#BFBFBF',
  DEFAULT = '#FFBB96',
}

export const getNotchColor = (spanType: SemanticGroupNames) => {
  switch (spanType) {
    case SemanticGroupNames.Http: {
      return NotchColor.HTTP;
    }

    case SemanticGroupNames.Database: {
      return NotchColor.DB;
    }

    case SemanticGroupNames.Rpc: {
      return NotchColor.RPC;
    }

    case SemanticGroupNames.Messaging: {
      return NotchColor.MESSAGING;
    }
    default: {
      return NotchColor.DEFAULT;
    }
  }
};

const getTextColor = (spanType: SemanticGroupNames) => {
  switch (spanType) {
    default: {
      return 'inherit';
    }
  }
};

export const TraceNode = styled.div<{selected: boolean}>`
  background-color: white;
  border: 1px solid ${({selected}) => (selected ? '#48586C' : '#E2E4E6')};
  border-radius: 2px;
  min-width: fit-content;
  display: flex;
  width: 150px;
  max-width: 150px;
  height: 75px;
  justify-content: center;
  align-items: center;
`;

export const TraceNotch = styled.div<{spanType: SemanticGroupNames}>`
  background-color: ${({spanType}) => getNotchColor(spanType)};
  position: absolute;
  top: 0px;
  margin-top: 1px;
  padding: 3px 6px;
  border-radius: 2px;
  border-bottom-left-radius: 0px;
  border-bottom-right-radius: 0px;
  width: 99%;
  font-weight: 700;
  align-items: center;

  span {
    color: ${({spanType}) => getTextColor(spanType)};
  }
`;

export const NameText = styled(Typography.Text).attrs({
  ellipsis: true,
})`
  margin: 0;
  font-size: 12px;
`;

export const TextContainer = styled.div`
  padding: 6px;
  padding-top: 30px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 150px;
  max-width: 150px;
`;
