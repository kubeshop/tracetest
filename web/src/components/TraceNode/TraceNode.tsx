import {NodeProps} from 'react-flow-renderer';
import {SemanticGroupNames} from '../../constants/SemanticGroupNames.constants';
import {ISpan} from '../../types/Span.types';
import GenericTraceNode from './components/GenericTraceNode';

export type TTraceNodeProps = NodeProps<ISpan>;

const ComponentMap: Record<string, typeof GenericTraceNode> = {
  [SemanticGroupNames.Http]: GenericTraceNode,
};

const TraceNode: React.FC<TTraceNodeProps> = ({data: span, ...props}) => {
  const Component = ComponentMap[span?.type || ''] || GenericTraceNode;

  return <Component data={span} {...props} />;
};

export default TraceNode;
