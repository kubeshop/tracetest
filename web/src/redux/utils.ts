import {PromiseWithKnownReason} from '@reduxjs/toolkit/dist/query/core/buildMiddleware/types';
import webSocketGateway, {IListenerFunction} from 'gateways/WebSocket.gateway';

export type {IListenerFunction} from 'gateways/WebSocket.gateway';

interface IInitWebSocketSubscription {
  listener: IListenerFunction;
  resource: string;
  waitToCleanSubscription: Promise<void>;
  waitToInitSubscription: PromiseWithKnownReason<any, any>;
}

/**
 * Implements RTK Query logic to receive streaming updates for persistent queries.
 * It allows a query to establish a WebSocket subscription, and apply updates to
 * the cached data when the information is received from the server.
 * For more information please take a look: https://redux-toolkit.js.org/rtk-query/usage/streaming-updates
 */
export async function initWebSocketSubscription({
  listener,
  resource,
  waitToCleanSubscription,
  waitToInitSubscription,
}: IInitWebSocketSubscription) {
  try {
    await waitToInitSubscription;
    webSocketGateway.subscribe(resource, listener);
  } catch {
    // no-op in case `waitToCleanSubscription` resolves before `waitToInitSubscription`,
    // in which case `waitToInitSubscription` will throw
  }
  await waitToCleanSubscription;
  webSocketGateway.unsubscribe(resource, '1234');
}
