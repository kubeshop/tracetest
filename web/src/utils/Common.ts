import {capitalize} from 'lodash';
import Env from './Env';

export const escapeString = (str: string): string => {
  // eslint-disable-next-line no-control-regex
  return str.replace(/[\\"']/g, '\\$&').replace(/\u0000/g, '\\0');
};

export const isBoolean = (value: string): boolean => value === 'true' || value === 'false';

export const isJson = (str: string) => {
  try {
    JSON.parse(str);
  } catch (e) {
    return false;
  }

  return Number.isNaN(Number(str)) && !isBoolean(str) && true;
};

export const getObjectIncludesText = (object: unknown, text: string = ''): boolean => {
  if (!text.length) return false;

  const searchTextLower = text.toLowerCase();
  const stringSpan = JSON.stringify(object).toLowerCase();

  return stringSpan.includes(searchTextLower);
};

export const downloadFile = (data: string, fileName: string): Element => {
  const element = document.createElement('a');
  const file = new Blob([data]);
  element.href = URL.createObjectURL(file);
  element.download = fileName;
  document.body.appendChild(element);
  element.click();
  return element;
};

export function singularOrPlural(noun: string, quantity: number) {
  if (quantity === 1) return noun;
  return `${noun}s`;
}

export function ordinalSuffixOf(i: number) {
  const j = i % 10;
  const k = i % 100;
  if (j === 1 && k !== 11) {
    return `${i}st`;
  }
  if (j === 2 && k !== 12) {
    return `${i}nd`;
  }
  if (j === 3 && k !== 13) {
    return `${i}rd`;
  }
  return `${i}th`;
}

export function getTotalCountFromHeaders(meta: any) {
  return Number(meta?.response?.headers.get('x-total-count') || 0);
}

export const getServerBaseUrl = () => {
  const {host, protocol} = window.location;
  const prefix = Env.get('serverPathPrefix');

  return `${protocol}//${host}${prefix}`;
};

export const ToTitle = (str: string) => {
  return capitalize(str.replace(/\W/g, ' '));
};

export const getIsValidUrl = (url: string): boolean => {
  try {
    return !!getParsedURL(url);
  } catch (e) {
    return false;
  }
};

export const getParsedURL = (rawUrl: string): URL => {
  if (!!rawUrl && !rawUrl.startsWith('http')) {
    return new URL(`http://${rawUrl}`);
  }

  return new URL(rawUrl);
};

export const withLowPriority =
  <T extends (...args: any[]) => any>(fn: T): ((...args: Parameters<T>) => Promise<ReturnType<T>>) =>
  (...args: Parameters<T>): Promise<ReturnType<T>> =>
    new Promise(resolve => {
      setTimeout(() => {
        resolve(fn(...args));
      }, 0);
    });

export const toPercent = (value: number) => {
  return `${value * 100}%`;
};

export const abbreviateNumber = (num: number): string => {
  const symbols = ['', 'k', 'M', 'G', 'T', 'P', 'E'];

  // tier to determine suffix
  // eslint-disable-next-line no-bitwise
  const tier = (Math.log10(num) / 3) | 0;

  // if zero not need a suffix
  if (tier === 0) return num.toString();

  // get suffix
  const suffix = symbols[tier];
  if (!suffix) return num.toString();

  // get scale
  const scale = 10 ** (tier * 3);
  const scaled = num / scale;
  const rounded = scaled.toFixed(1);

  return rounded + suffix;
};
