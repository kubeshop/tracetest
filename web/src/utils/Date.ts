import {format, formatDistanceToNowStrict, isValid, parseISO} from 'date-fns';
import dropWhile from 'lodash/dropWhile';
import round from 'lodash/round';

export const ONE_MILLISECOND = 1000 * 1;
const ONE_SECOND = 1000 * ONE_MILLISECOND;
const ONE_MINUTE = 60 * ONE_SECOND;
const ONE_HOUR = 60 * ONE_MINUTE;
const ONE_DAY = 24 * ONE_HOUR;

const UNIT_STEPS: {unit: string; microseconds: number; ofPrevious: number}[] = [
  {unit: 'd', microseconds: ONE_DAY, ofPrevious: 24},
  {unit: 'h', microseconds: ONE_HOUR, ofPrevious: 60},
  {unit: 'm', microseconds: ONE_MINUTE, ofPrevious: 60},
  {unit: 's', microseconds: ONE_SECOND, ofPrevious: 1000},
  {unit: 'ms', microseconds: ONE_MILLISECOND, ofPrevious: 1000},
  {unit: 'Î¼s', microseconds: 1, ofPrevious: 1000},
];

const Date = {
  format(date: string, dateFormat = "EEEE, yyyy/MM/dd 'at' HH:mm:ss") {
    const isoDate = parseISO(date);
    if (!isValid(isoDate)) {
      return '';
    }
    return format(isoDate, dateFormat);
  },

  getTimeAgo(date: string) {
    const isoDate = parseISO(date);
    if (!isValid(isoDate)) {
      return '';
    }
    return formatDistanceToNowStrict(isoDate, {addSuffix: true});
  },

  isDefaultDate(date: string) {
    return date === '0001-01-01T00:00:00Z';
  },

  /**
   * Format duration for display.
   *
   * @param {number} duration - microseconds
   * @return {string} formatted duration
   */
  formatDuration(duration: number): string {
    // Drop all units that are too large except the last one
    const [primaryUnit, secondaryUnit] = dropWhile(
      UNIT_STEPS,
      ({microseconds}, index) => index < UNIT_STEPS.length - 1 && microseconds > duration
    );

    if (primaryUnit.ofPrevious === 1000) {
      // If the unit is decimal based, display as a decimal
      return `${round(duration / primaryUnit.microseconds, 2)}${primaryUnit.unit}`;
    }

    const primaryValue = Math.floor(duration / primaryUnit.microseconds);
    const primaryUnitString = `${primaryValue}${primaryUnit.unit}`;
    const secondaryValue = Math.round((duration / secondaryUnit.microseconds) % primaryUnit.ofPrevious);
    const secondaryUnitString = `${secondaryValue}${secondaryUnit.unit}`;
    return secondaryValue === 0 ? primaryUnitString : `${primaryUnitString} ${secondaryUnitString}`;
  },
};

export default Date;
