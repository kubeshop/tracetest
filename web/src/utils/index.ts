import {COMPARE_OPERATOR} from 'types';

const flattenAttributesSelector = () =>
  `resourceSpans[].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0]},resource.attributes[].{key:key,value: value.*|[0]}]|[][]`;

export const filterBySpanId = (spanId: string = '') =>
  `resourceSpans[?instrumentationLibrarySpans[?spans[?starts_with(spanId,'${spanId}')]]] | [].[instrumentationLibrarySpans[].spans[].attributes[].{key:key,value:value.*|[0],type:'span'},resource.attributes[].{key:key,value: value.*|[0],type:'resource'}]|[][]`;

export const getOperator = (op: COMPARE_OPERATOR) => {
  switch (op) {
    case COMPARE_OPERATOR.EQUALS:
      return 'eq';
    case COMPARE_OPERATOR.NOTEQUALS:
      return 'ne';
    case COMPARE_OPERATOR.GREATERTHAN:
      return 'gt';
    case COMPARE_OPERATOR.LESSTHAN:
      return 'lt';
    case COMPARE_OPERATOR.GREATOREQUALS:
      return 'gte';
    case COMPARE_OPERATOR.LESSOREQUAL:
      return 'lte';
    default:
      return 'eq';
  }
};
