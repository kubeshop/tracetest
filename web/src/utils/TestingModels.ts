import {HTTP_METHOD} from '../constants/Common.constants';
import {CompareOperator} from '../constants/Operator.constants';
import {SemanticGroupNames} from '../constants/SemanticGroupNames.constants';
import {LOCATION_NAME} from '../constants/Span.constants';
import {SpanAttributeType} from '../constants/SpanAttribute.constants';
import {TestState} from '../constants/TestRunResult.constants';
import {IAssertion, IAssertionResult, IItemSelector, ISpanAssertionResult} from '../types/Assertion.types';
import {ISpan} from '../types/Span.types';
import {ITest} from '../types/Test.types';
import {ITestRunResult} from '../types/TestRunResult.types';
import {ITrace} from '../types/Trace.types';

interface IProps {
  mouseEvent: MouseEvent;
  assertionId: string;
  spanAssertionResult: ISpanAssertionResult;
  trace: ITrace;
  span: ISpan;
  test: ITest;
  assertion: IAssertion;
  assertionResult: IAssertionResult;
  testRunResult: ITestRunResult;
  resultId: string;
  testId: string;
  spanId: string;
}

const resultId = '0932u0e3';
const testId = '2309ei30';
const spanId = '4938u43';

const test: ITest = {
  assertions: [],
  description: '',
  lastTestResult: undefined,
  name: '',
  serviceUnderTest: {
    id: '',
    request: {
      auth: undefined,
      body: '',
      certificate: undefined,
      headers: undefined,
      method: HTTP_METHOD.GET,
      proxy: undefined,
      url: '',
    },
  },
  testId,
};

const span: ISpan = {
  attributeList: [],
  attributes: {},
  duration: 0,
  endTimeUnixNano: '',
  kind: '',
  name: '',
  parentSpanId: '',
  signature: [
    {
      locationName: LOCATION_NAME.SPAN,
      propertyName: '',
      value: '',
      valueType: '',
    },
  ],
  spanId,
  startTimeUnixNano: '',
  status: {code: ''},
  traceId: '',
  type: SemanticGroupNames.Http,
};

let selector: IItemSelector = {
  locationName: LOCATION_NAME.SPAN,
  propertyName: '',
  value: '',
  valueType: '',
};

const assertionResult: IAssertionResult = {
  assertion: {
    assertionId: '',
    selectors: [selector],
    spanAssertions: [],
  },
  spanListAssertionResult: [
    {
      resultList: [
        {
          actualValue: 'New request',
          comparisonValue: '',
          hasPassed: false,
          locationName: LOCATION_NAME.SPAN,
          operator: CompareOperator.EQUALS,
          propertyName: '',
          spanAssertionId: '',
          spanId: '',
          valueType: SpanAttributeType.intValue,
        },
      ],
      span,
    },
  ],
};

const testRunResult: ITestRunResult = {
  executionTime: 0,
  failedAssertionCount: 0,
  passedAssertionCount: 0,
  totalAssertionCount: 0,

  assertionResult: [{assertionId: '', spanAssertionResults: []}],
  assertionResultState: false,
  completedAt: '',
  createdAt: new Date().toISOString().toString(),
  response: undefined,
  resultId: '',
  spanId: '',
  state: TestState.CREATED,
  testId: '',
  trace: {
    description: '',
    resourceSpans: undefined,
    spans: [span],
  },
  traceId: '',
};

const spanAssertionResult: ISpanAssertionResult = {
  actualValue: 'New request',
  comparisonValue: '',
  hasPassed: false,
  locationName: LOCATION_NAME.INSTRUMENTATION_LIBRARY,
  operator: CompareOperator.EQUALS,
  propertyName: '',
  spanAssertionId: '',
  spanId: '',
  valueType: SpanAttributeType.doubleValue,
};

const assertionId = '234234';

const assertion: IAssertion = {
  assertionId,
  selectors: [
    {
      locationName: LOCATION_NAME.INSTRUMENTATION_LIBRARY,
      propertyName: '',
      value: '',
      valueType: '',
    },
  ],
  spanAssertions: [],
};
export const TestingModels: IProps = {
  mouseEvent: new MouseEvent('click', {
    bubbles: true,
    cancelable: true,
  }),
  assertionId,
  assertion,
  spanAssertionResult,
  assertionResult,
  test,
  testRunResult,
  span,
  resultId,
  testId,
  spanId,
  trace: {
    description: '',
    spans: [span],
  },
};

class NewTestingModel {
  test: ITest = test;
  assertionResult: IAssertionResult = assertionResult;
  testRunResult: ITestRunResult = testRunResult;
  span: ISpan = span;
  resultId: string = resultId;
  testId: string = testId;
  spanId: string = spanId;
  trace: ITrace = {
    description: '',
    spans: [span],
  };
}

export const TestingObj = new NewTestingModel();
