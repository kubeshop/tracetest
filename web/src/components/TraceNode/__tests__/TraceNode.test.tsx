import {render} from '@testing-library/react';
import {ReactFlowProvider} from 'react-flow-renderer';
import {TestingModels} from '../../../utils/TestingModels';
import TraceNode from '../TraceNode';

test('TraceNode', () => {
  const result = render(
    <ReactFlowProvider>
      <TraceNode
        id="sdfkjn"
        type="ksdjfn"
        data={TestingModels.span}
        dragHandle="sdkjf"
        isDragging={false}
        selected={false}
        isConnectable={false}
      />
    </ReactFlowProvider>
  );
  expect(result.container).toMatchSnapshot();
});
