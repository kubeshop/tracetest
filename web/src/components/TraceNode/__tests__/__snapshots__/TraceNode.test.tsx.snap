// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceNode 1`] = `
<div>
  <div
    class="sc-bdvvtL kNAymc"
    data-cy="trace-node-http"
  >
    <div
      class="sc-gsDKAQ cmaxSH"
    >
      <span
        class="ant-typography"
      >
        HTTP
      </span>
    </div>
    <div
      class="react-flow__handle react-flow__handle-top nodrag target connectable"
      data-handleid="sdfkjn"
      data-handlepos="top"
      style="top: 0px; border-radius: 0; visibility: hidden;"
    />
    <div
      class="sc-hKwDye etMbsh"
    >
      
      <span
        class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line sc-dkPtRN dvhzfk"
      >
        
      </span>
    </div>
    <div
      class="react-flow__handle react-flow__handle-bottom nodrag source connectable"
      data-handleid="sdfkjn"
      data-handlepos="bottom"
      style="bottom: 0px; border-radius: 0; visibility: hidden;"
    />
  </div>
</div>
`;
